gmail-to-salesforce-lead/main.bal

```diff
@@ -0,0 +1,193 @@
+    removeLabels(gmailClient, matchingMailThreads, labelIdsToMatch);
+
+    gmail:Message[] matchingEmails = getMatchingEmails(gmailClient, matchingMailThreads);
+
+    Email[] emails = [];
+    from gmail:Message message in matchingEmails
+    do {
+        Email|error email = parseEmail(message);
+        if email is Email {
+            emails.push(email);
+        }    
+    };
+    
+    return emails;
+}
+
+function getLabelIds(gmail:Client gmailClient, string[] labelsToMatch) returns string[]|error {
+    gmail:LabelList labelList = check gmailClient->listLabels("me");
+
+    return from gmail:Label label in labelList.labels
+        where labelsToMatch.indexOf(label.name) != ()
+        select label.id;
+}
+
+function getMatchingMailThreads(gmail:Client gmailClient, string[] labelIdsToMatch) returns gmail:MailThread[]|error {
+    gmail:MsgSearchFilter searchFilter = {
+        includeSpamTrash: false,
+        labelIds: labelIdsToMatch
+    };
+
+    stream<gmail:MailThread, error?>|error mailThreadStream = gmailClient->listThreads(filter = searchFilter);
+    if mailThreadStream is error {
+        return mailThreadStream;
+    }
+
+    return check from gmail:MailThread mailThread in mailThreadStream
+        select mailThread;
+}
+
+function removeLabels(gmail:Client gmailClient, gmail:MailThread[] mailThreads, string[] labelIds) {
+    from gmail:MailThread mailThread in mailThreads
+    do {
+        gmail:MailThread|error removeLabelResponse = gmailClient->modifyThread(mailThread.id, [], labelIds);
+        if removeLabelResponse is error {
+            log:printError("An error occured in removing the labels from the thread.", removeLabelResponse, removeLabelResponse.stackTrace(), threadId = mailThread.id, labelIds = labelIds);
+        }
+    };
+}
+
+function getMatchingEmails(gmail:Client gmailClient, gmail:MailThread[] mailThreads) returns gmail:Message[] {
+    gmail:Message[] messages = [];
+    _ = from gmail:MailThread mailThread in mailThreads
+        do {
+            gmail:MailThread|error response = gmailClient->readThread(mailThread.id);
+            if response is error {
+                log:printError("An error occured while reading the email.", response, response.stackTrace(), threadId = mailThread.id);
+            } else {
+                messages.push((<gmail:Message[]>response.messages)[0]);
+            }
+        };
+
+    return messages;
+}
+
+function parseEmail(gmail:Message message) returns Email|error {
+    string 'from = <string>message.headerFrom;
+    string subject = <string>message.headerSubject;
+    string body = <string>(check mime:base64Decode(<string>(<gmail:MessageBodyPart>message.emailBodyInText).data));
+
+    return {
+        'from: 'from,
+        subject: subject,
+        body: body
+    };
+}
+
+function generateLead(string 'from, string subject, string body) returns Lead|error {
+    openAI:Client openAIClient = check new ({
```

Comment:
Seems this client is also created during each iteration of loop.

Reply:
Fixed in [cce2eaf](https://github.com/ballerina-guides/integration-samples/pull/32/commits/cce2eafff5580b9fc4de69c431a8951e6c90b36a)